### PR TITLE
Feature/chapter 07

### PR DESCRIPTION
**Pull request by jeroenvs**

### CODE REVIEW

File: 07-documentation.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+Represents a person that uses our system. Each user has a role, which is also a seperate resource.
+
+- Get a user
+- Create user
+- Update user
+- Remove user
+
+**Get a user**
+
+	GET /api/user/1
```
Comment (SanderBenschop) [inline 07-documentation.md]: In chapter two we describe that we use plurals for resources. Can you change the example to reflect that? :)

**Comment by jeroenvs** [comment]
Done

File: 07-documentation.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
+# REST documentation
+
+With the separation of frontend and backend it has become of critical importance that the REST interface is clearly documented. It is possible that frontend and backend tasks are performed by different developers, which simply means that the frontender should know what requests to call and what output to expect.
```
Comment (Pouja) [inline 07-documentation.md]: `frontender` is not a word: frontend developer would be better.

Comment (Pouja) [inline 07-documentation.md]: I think it is `which requests to call and what kind of output the developer should expect.`

File: 07-documentation.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+[Swagger](http://swagger.io/) provides a Spring MVC addon that inspects all REST endpoints and generates an interface description. In combination with [Swagger UI](http://swagger.io/swagger-ui/) a visual test suite can be rendered, showing each controller and request. The requests can be invoked on demand, with parameter and request body support.
+
+Additional documentation can be added by enhancing the controller methods with annotations:
+
+	@ApiResponses(value = {@ApiResponse(code = 405, message = "Invalid input")})
+	public .... createSomething(..)
+
+Swagger is easy to integrate in projects but also has various drawbacks:
+
+- *No cross cutting concerns.* Swagger only shows the requests, there is no way to define cross cutting concerns. One clear example is the usage of error codes. Each request shows a list of all possible error codes, most are clearly irrelevant for that request, but could confuse a frontend developer. The only way to overcome this problem is by heavily annotating your controller with Swagger annotations.
```
Comment (Pouja) [inline 07-documentation.md:24]: Or you can include all the errors in the documentation of the method.

File: 07-documentation.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
+# REST documentation
+
+With the separation of frontend and backend it has become of critical importance that the REST interface is clearly documented. It is possible that frontend and backend tasks are performed by different developers, which simply means that the frontender should know what requests to call and what output to expect.
+
+Besides an overview of resources the documentation should also discuss cross cutting concerns, such as status codes, error handling and security. The REST document serves as contract, helping consumers of the interface and ensuring that the interface stays consistent and clear.
```
Comment (Pouja) [inline 07-documentation.md:5]: It is still not clear for me what you mean with `cross cutting concerns` as you use it in several different context over the whole document.

File: 07-documentation.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+- *Grouping per controller not resource.* Sometimes a resource is split over various controllers for design reasons. Each controller is shown seperately, while you would rather see them as one resource.
+
+## Spring REST docs
+
+[Spring REST docs](http://docs.spring.io/spring-restdocs/docs/1.0.x/reference/html5/) helps to generate an acurate and readable REST document. Rather than a testsuite we will produce an actual document, using the text processor [Ascii doctor](http://asciidoctor.org/). In this document we can manually describe the cross cutting concerns and resources, allowing us to manage our own markup and chapters.
+
+Frequently changing information, such as HTTP request details, are generated. Spring REST docs generates various snippets from unit tests written in [Spring MVC Test](http://docs.spring.io/spring/docs/current/spring-framework-reference/htmlsingle/#spring-mvc-test-framework). By using unit tests you ensure that the documentation is actual, otherwise the test will fail. Another nice side effect is that each request will be tested.
```
Comment (Pouja) [inline 07-documentation.md:31]: are generated by who, what, where, why? What do you mean with the first sentence?

Comment (Pouja) [inline 07-documentation.md:31]: is `actual`?

Comment (Pouja) [inline 07-documentation.md:31]: I am confused with this paragraph, you are talking about generating rest documentation, but here you are talking about testing? Is the documentation generated based on the unit tests you write?

File: 07-documentation.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+With the separation of frontend and backend it has become of critical importance that the REST interface is clearly documented. It is possible that frontend and backend tasks are performed by different developers, which simply means that the frontender should know what requests to call and what output to expect.
+
+Besides an overview of resources the documentation should also discuss cross cutting concerns, such as status codes, error handling and security. The REST document serves as contract, helping consumers of the interface and ensuring that the interface stays consistent and clear.
+
+Writing REST documention takes time, but should be worth it in the process of a project. REST documentation minimizes communication issues between developers and helps new developers become productive in a shorter time span. To really benefit from a REST document it has to stay accurate. Outdated documentation can bring more problems than solutions.
+
+# Tooling
+
+We use tooling to write and maintain the REST documentation in a minimal amount of time. In the sections below two tools are described and compared. Use the tool that best fits your situation.
```
Comment (Pouja) [inline 07-documentation.md:11]: I am not really found of using first person pov in a technical document. But what do you guys think @42BV/developer ?

File: 07-documentation.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+## Example resource
+
+To make this clearer we will provide an example resource description below.
+
+### User
+
+Represents a person that uses our system. Each user has a role, which is also a seperate resource.
+
+- Get a user
+- Create user
```
Comment (Pouja) [inline 07-documentation.md]: Either do not mention create, update and remove user or include them in the example. It is kinda confusing.

File: 07-documentation.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+	{
+		id: 1,
+		email: developer@42.nl,
+		role: {
+			name: "ADMIN"
+		}
+	}
+
+Another nice example is [Github](https://developer.github.com/v3/git/commits/). Where they describe their *Commit* resource.
```
Comment (Pouja) [inline 07-documentation.md]: Maybe change it in something more like: `A good example of a REST documentation is [this](https://developer.github.com/v3/git/commits/)  where GitHub describes their *commit* resource of their REST api.`

And maybe place under a separate header? Now it looks like it is part of the user example resource.

File: 07-documentation.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
+# REST documentation
+
```
Comment (robert-bor) [inline 07-documentation.md:2]: Will this entire chapter be impacted by your investigation of next-gen rest documentation tools?

File: 07-documentation.md (hunk 4 above)
Comment (robert-bor) [inline 07-documentation.md:5]: Cross-cutting concerns are technical aspects, which are required for several functional methods. Typically aspect-oriented programming is used to insert this logic. A couple of examples of cross-cutting concerns are writing log statements about the duration of a method call, or the conversion of a NotFoundException to a 404 error. These cross-cutting concerns are best handled in a generic way.

File: 07-documentation.md (hunk 6 above)
Comment (robert-bor) [inline 07-documentation.md:11]: Best use the imperative, or advising form.

File: 07-documentation.md (hunk 3 above)
Comment (robert-bor) [inline 07-documentation.md:24]: With the disadvantage that they will have to be added over and over again for all methods that have that error.

File: 07-documentation.md (hunk 5 above)
Comment (robert-bor) [inline 07-documentation.md:31]: Afaik, and from what Jeroen showed me, it's a two-step process. The unit tests generate request/response pairs, which are then used by the REST documentation generator to generate the site.

**Pull request merged**